### PR TITLE
Fix TensorList constructor from list of tensors.

### PR DESCRIPTION
**Pull request by ksztenderski**
## Description
- [x] **Bug fix** (*non-breaking change which fixes an issue*)
- [ ] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR

It fixes a bug with TensorList constructor from list of tensors.
Fixed issues (raising exceptions), when constructing TensorList from:
- an empty list
- different backend tensors
- different dtype tensors

#### Additional information
- Affected modules and functionalities:

TensorList constructor from list of Tensors

- Key points relevant for the review:

### Tests
- [x] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by ksztenderski** [comment]
!build

File: dali/python/backend_impl.cc [hunk 1]
```diff
@@ -596,11 +596,34 @@ std::unique_ptr<Tensor<Backend> > TensorListGetItemImpl(TensorList<Backend> &t,
...
+        expected_type = t.type();
+      } else if (expected_type != cur_type) {
+        throw py::type_error(make_string(
+            "Tensors cannot have different data types. Tensor at position ", i, " has type '",
+            cur_type, "' expected to have type '", DALIDataType(expected_type), "'."));
+      }
+
+      tv[i].ShareData(t);
+    } catch (const py::type_error &e) {
+      throw e;
```
Comment (prak-nv) [inline dali/python/backend_impl.cc:622]: Nitpick, just `throw;` would have rethrow `e`.

Comment (ksztenderski) [reply dali/python/backend_impl.cc]: done

File: dali/python/backend_impl.cc [hunk 2]
```diff
@@ -596,11 +596,34 @@ std::unique_ptr<Tensor<Backend> > TensorListGetItemImpl(TensorList<Backend> &t,
 template <typename Backend>
 std::shared_ptr<TensorList<Backend>> TensorListFromListOfTensors(py::list &list_of_tensors,
                                                                  string &layout) {
+  if (list_of_tensors.empty()) {
+    throw std::runtime_error("Cannot create TensorList from an empty list.");
+  }
+
   auto tl = std::make_shared<TensorList<Backend>>(list_of_tensors.size());
   TensorVector<Backend> tv(list_of_tensors.size());
+  int expected_type = -2;
```
Comment (prak-nv) [inline dali/python/backend_impl.cc:605]: Why DALI_NO_TYPE ie. `-1` wouldn't work here?

Comment (ksztenderski) [reply dali/python/backend_impl.cc:605]: It seems like a better (cleaner) choice, but I think that it is theoretically possible for a Tensor to be of type `DALI_NO_TYPE`, but maybe I am overcomplicating things.

**Review by prak-nv — APPROVED** [review]

**Review by klecki — APPROVED** [review]

**Comment by ksztenderski** [comment]
!build

**Pull request merged**